### PR TITLE
Add Logistic Regression to our NLU classifiers.

### PR DESCRIPTION
**Pull request by koaning**
This PR will add a logistic regression. The related issue is https://github.com/RasaHQ/rasa/issues/10649.

**Status (please check what you already did)**:

- [x] added some tests for the functionality
- [x] ran the benchmarks to confirm the merits of this component
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)

### CODE REVIEW

File: rasa/nlu/classifiers/logistic_regression_classifier.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    def __init__(
+        self,
+        config: Dict[Text, Any],
+        name: Text,
+        model_storage: ModelStorage,
+        resource: Resource,
+    ) -> None:
+        """Construct a new classifier."""
+        self.name = name
+        self.clf = LogisticRegression(
```
Comment (mleimeister) [inline rasa/nlu/classifiers/logistic_regression_classifier.py:53]: Would we want to expose more parameters than these? E.g. `random_state` would be useful for reproducibility. `solver` seems a very technical detail, and afaik there's some cross-dependency with the `penalty. Should in this case both be configurable, so users can also chose `l1` or `elasticnet`? Alternatively, could also pass on `**config` filtered for all valid parameter choices?

Comment (koaning) [reply rasa/nlu/classifiers/logistic_regression_classifier.py:53]: Oh totally! The current PR is a bit more work in progress than "ready for review". I'll update the title to reflect this. 

At the moment I'm still running the benchmarks so that we can confirm that it's not an unreasonable alternative. 

In general, I think it's best to be explicit in the parameters that we allow. The **kwargs trick is nice for quick iteration, but requires deep knowledge of sklearn which can get in the way of using the component.

**Comment by koernerfelicia** [comment]
> The current PR is a bit more work in progress than "ready for review"

@koaning I hadn't realised that and came here feeling a bit bad I was late to give you a review 😅 Would you, please, ping me when it *is* ready? Alternatively, @mleimeister can also continue to review when you're ready. Or we both can, if you'd like multiple eyes and Matthias is fine with that. Whatever works, just let me know!

**Comment by koaning** [comment]
@koernerfelicia my bad! This was a miscommunication on my part. I figured I'd make the PR but then Daksh (quite reasonably) asked if I could run a few benchmarks first just to be sure that this is a reasonable addition. The results are currently being discussed on [slack](https://rasa-hq.slack.com/archives/C8LEHMTAM/p1642007499069100).

**Comment by koernerfelicia** [comment]
@koaning no worries! We've discussed before that the auto-tagging for code-owners is not an ideal system. Will definitely be bringing this up in the next research retro. I'll check out the Slack thread!

File: rasa/nlu/classifiers/logistic_regression_classifier.py (hunk 1 above)
Comment (koaning) [reply rasa/nlu/classifiers/logistic_regression_classifier.py:53]: I've just added the random seed but I'm wondering if it makes sense to either be a bit minimal in the parameters that we allow or if we want to allow the user to fully customize everything. 

@koernerfelicia @mleimeister got any strong opinions on this?

**Comment by koaning** [comment]
I've removed the WIP flag @koernerfelicia @mleimeister. I'm still working on the docs, but I think the code/tests are at a stage where I'm certainly open to suggestions.

File: rasa/nlu/classifiers/logistic_regression_classifier.py (hunk 1 above)
Comment (mleimeister) [reply rasa/nlu/classifiers/logistic_regression_classifier.py:59]: Apart from allowing full customization (which I think we don't necessarily need), another argument for being explicit about all default parameters is transparency and ensuring reproducibility. If e.g. Sklearn would change a default, we wouldn't notice, but the component suddenly behaves differently. I think we list all defaults like this in `CountVectorsFeaturizer`, but some would rarely be set in the pipeline config.

File: rasa/nlu/classifiers/logistic_regression_classifier.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        examples = [
+            e
+            for e in training_data.intent_examples
+            if (e.get("intent") and e.get("text"))
+        ]
+
+        for e in examples:
+            # First element is sequence features, second is sentence features
+            # for intent classification we're only interested in sentence features
+            sparse_feats = e.get_sparse_features(attribute=TEXT)[1]
```
Comment (mleimeister) [inline rasa/nlu/classifiers/logistic_regression_classifier.py:102]: The extraction of `X` from an example looks like the same as in `_create_X`. Could this be merged into a helper function to be re-used in both places?

File: rasa/nlu/classifiers/logistic_regression_classifier.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
+import logging
+from typing import Any, Text, Dict, List, Type
+
+from joblib import dump, load
```
Comment (mleimeister) [inline rasa/nlu/classifiers/logistic_regression_classifier.py:4]: Since `dump` and `load` could be occurring in other modules as well and lead to shadowing, it might be better to call them by their full module path, as per [here](https://www.notion.so/rasa/Python-Code-Conventions-26835ffd72484dd699f28fa693916d2e#d5fb5c6142394ff99505dc8160255ef5):

```
import joblib
joblib.dump(...)
```

File: rasa/nlu/classifiers/logistic_regression_classifier.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    @classmethod
+    def load(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> GraphComponent:
+        """Loads trained component (see parent class for full docstring)."""
+        with model_storage.read_from(resource) as model_dir:
```
Comment (mleimeister) [inline rasa/nlu/classifiers/logistic_regression_classifier.py:174]: To cover the case where there is no persisted data for the component, this should be covered in an exception block catching `ValueError`.

Comment (koaning) [reply rasa/nlu/classifiers/logistic_regression_classifier.py:174]: Is there an example of this? When would there not be any persisted data?

Comment (mleimeister) [reply rasa/nlu/classifiers/logistic_regression_classifier.py:184]: [This](https://github.com/RasaHQ/rasa/blob/main/rasa/nlu/classifiers/diet_classifier.py#L1074) would be an example. I don't think this would ever happen programatically, but I think if e.g. someone deleted the individual cached model folder in `.rasa`, but not the database containing the pointers to the cached components, the lookup would fail.

Comment (koaning) [reply rasa/nlu/classifiers/logistic_regression_classifier.py:184]: This begs the question: do we have a unit test for this behavior?

**Comment by koaning** [comment]
@mleimeister what should I do with the codeclimate warning by the way? Is that something that we should ignore? It'll be hard to remove complexity there.

**Comment by mleimeister** [comment]
> @mleimeister what should I do with the codeclimate warning by the way? Is that something that we should ignore? It'll be hard to remove complexity there.

I think that's fine as it's not a required check. But seems to have lifted itself :)

File: tests/nlu/classifiers/test_logistic_regression_classifier.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    # Make predictions.
+    classifier.process(training_data.training_examples)
+
+    # Check that the messages have been processed correctly
+    for msg in training_data.training_examples:
+        _, conf = msg.get("intent")["name"], msg.get("intent")["confidence"]
+        # Confidence should be between 0 and 1.
+        assert 0 < conf
+        assert conf < 1
```
Comment (mleimeister) [inline tests/nlu/classifiers/test_logistic_regression_classifier.py:84]: Can move to single line: `assert 0 < conf < 1`

**Review by mleimeister — COMMENTED** [review]
For unit tests, in order to cover the full functionality, the following should be added:
- Train -> persist -> load -> predict, as e.g. [here](https://github.com/RasaHQ/rasa/blob/main/tests/nlu/classifiers/test_keyword_classifier.py#L38)
- Test with featurizers using only dense / sparse+dense features (dense could be just some dummy vectors)
- Include `LogisticRegressionClassifier` in `tests.nlu.test_*`
- Empty text messages (if this is expected to happen?)

**Comment by koaning** [comment]
@mleimeister could you clarify what you mean with `tests.nlu.test_*`? I've added an element to the `test_train.py` file. Is that what you meant?

**Comment by koaning** [comment]
@mleimeister I'm getting close. One question on the empty message ... what's the expected behavior? Simply that we don't break?

File: docs/docs/components.mdx [hunk 6]
```diff
@@ -1052,6 +1052,83 @@ Intent classifiers assign one of the intents defined in the domain file to incom
   ```
 
 
+### LogisticRegressionClassifier
+
+* **Short**
+
+  Logistic regression intent classifier, using the [scikit-learn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).
```
Comment (mleimeister) [inline docs/docs/components.mdx:1059]: Not sure how pedantic we are about external links, but should this be hard-coded to the currently used version (I think it's `0.24.2`?) in case sklearn changes its implementation in future versions. On the other hand, it's quite likely that this link will not get updated if we upgrade the sklearn version either. So no real preference here 😄

File: rasa/nlu/classifiers/logistic_regression_classifier.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        probas = self.clf.predict_proba(X)
+        for idx, message in enumerate(messages):
+            intent = {"name": pred[idx], "confidence": probas[idx].max()}
+            intents = self.clf.classes_
+            intent_info = {
+                k: v
+                for i, (k, v) in enumerate(zip(intents, probas[idx]))
+                if i < LABEL_RANKING_LENGTH
+            }
+            intent_ranking = [
```
Comment (mleimeister) [inline rasa/nlu/classifiers/logistic_regression_classifier.py:154]: One issue I noticed with using this component with Moodbot is that the intent ranking seems to be unsorted. For a user text `hello`, I got the following scores:

```
[{'name': 'affirm', 'confidence': 0.018605408851111678}, {'name': 'bot_challenge', 'confidence': 0.0006403107430143258}, {'name': 'deny', 'confidence': 0.025455037690681255}, {'name': 'goodbye', 'confidence': 0.013747843491355998}, {'name': 'greet', 'confidence': 0.914782164381611}, {'name': 'mood_great', 'confidence': 0.017658700205053904}, {'name': 'mood_unhappy', 'confidence': 0.009110534637171744}]
```

This in turn leads to `FallbackClassifier` being triggered, since it will look at the difference of the first two elements, which is below the default ambiguity threshold. Therefore the probabilities need to be sorted, as e.g. [here](https://github.com/RasaHQ/rasa/blob/main/rasa/nlu/classifiers/sklearn_intent_classifier.py#L257).

Comment (koaning) [reply rasa/nlu/classifiers/logistic_regression_classifier.py:154]: Well spotted!

File: docs/docs/components.mdx [hunk 8]
```diff
@@ -1052,6 +1052,83 @@ Intent classifiers assign one of the intents defined in the domain file to incom
...
+        {
+            "confidence": 0.780,
+            "name": "greet"
+        },
+        {
+            "confidence": 0.140,
+            "name": "goodbye"
+        },
+        {
+            "confidence": 0.0810,
```
[the remark this answers is not in the document]
Comment (koaning) [reply docs/docs/components.mdx:1090]: Ah that shouldn't happen. Lemme fix that!

File: docs/docs/components.mdx (hunk 6 above)
Comment (koaning) [reply docs/docs/components.mdx:1059]: I think it's best to leave as-is. Mainly because keeping an eye on the docs is likely going to be more work and it's not unlikely that Rasa will keep sklearn updated.

File: rasa/nlu/classifiers/logistic_regression_classifier.py (hunk 7 above)
Comment (koaning) [reply rasa/nlu/classifiers/logistic_regression_classifier.py:154]: Also added a test for it.

**Comment by mleimeister** [comment]
> @mleimeister I'm getting close. One question on the empty message ... what's the expected behavior? Simply that we don't break?

Yes I think that's what some of the other components do in terms of checks, e.g. [here](https://github.com/RasaHQ/rasa/blob/main/rasa/nlu/classifiers/sklearn_intent_classifier.py#L198). 

However, I was not able to break the component with empty text. It moreover seems that even with empty text, our featurizers do produce feature vectors for the `Message`, so they are actually never empty when they arrive at the classifier (see screenshot below). Therefore I'd say the current behaviour is good as it is and an upcoming rework of the NLU components should rather define the contract between different components more clearly, after which it would be much easier to say what such checks should look like.

In that sense LGMT 👍 

Screenshot 2022-02-09 at 16 45 18

**Review by mleimeister — APPROVED** [review]

**Comment by koaning** [comment]
🎉
